fpmsyncd/fpmsyncd.cpp

```diff
@@ -18,25 +29,70 @@ int main(int argc, char **argv)
     NetDispatcher::getInstance().registerMessageHandler(RTM_NEWROUTE, &sync);
     NetDispatcher::getInstance().registerMessageHandler(RTM_DELROUTE, &sync);
 
-    while (1)
+    while (true)
     {
         try
         {
             FpmLink fpm;
             Select s;
 
-            cout << "Waiting for connection..." << endl;
+            SelectableTimer warmStartTimer(timespec{0, 0});
+
+            cout << "Waiting for fpm-client connection..." << endl;
             fpm.accept();
             cout << "Connected!" << endl;
 
             s.addSelectable(&fpm);
+
+            /* Initialize warm-restart logic if this one is enabled */
+            bool warmStartEnabled = sync.m_warmStartHelper.isEnabled();
+            if (warmStartEnabled)
+            {
+                /* Obtain warm-restart timer defined for routing application */
+                uint32_t warmRestartIval = sync.m_warmStartHelper.getRestartTimer();
+                if (!warmRestartIval)
+                {
+                    warmStartTimer.setInterval(timespec{DEFAULT_ROUTING_RESTART_INTERVAL, 0});
+                }
+                else
+                {
+                    warmStartTimer.setInterval(timespec{warmRestartIval, 0});
+                }
+
```

Comment:
I have concern on assigning a static value to the warmStartTimer, because it's hard for an admin to identify a good number for such timer:
-  if it's too short, obviously some routes will be missing.
-  if it's too long, black hole may occur, as shown in the following example:  a new prefix is advertised to BGP peers,  peers have the route/path installed immediately and start to forward traffic to the warm rebooted box. However, the prefix is NOT yet installed in warm rebooted box's forwarding table due to the  running warmStartTimer (not expire yet), hence packet drop.  It looks worse, if the peers do have other paths to reach the prefix, which is true in most of DC setup with ECMP.  I think loop may also occur, as long as the rebooted box's control plane is inconsistent with its forwarding plane.  The longer the inconsistency, the higher chance the problems happen.
In order to reduce the chance, the forwarding path needs to coordinate with routing/zebra stack  and set timer value accordingly.

Reply:
@heidinet2007 absolutely, we covered these cases that you are describing in my design doc (find link below -- search for Known-Limitations section and also look at the CLI portion).

As mentioned in this doc, for the time being user will need to set both timers (bgp and warm-restart one) to similar/same values to avoid inconsistencies. And yes, i'm already working in an FRR extension that will allow bgp/zebra to notify fpmsyncd whenever all BGP EoR's are received. That should address all these issues.

https://github.com/Azure/SONiC/blob/dcac72377f23521a394694214678ea4450f6f70a/doc/routing-warm-reboot/Routing_Warm_Reboot.md)

Reply:
It would be hard to deploy routing warm reboot without the auto sync between zebra and fpmsyncd (even though  the update-delay timer and warmstarttimer can be tuned),  because admins have alternative ways to achieve zero packet loss during routing stack upgrade.   When will the auto sync be done? Is it possible to include it in this release?

Reply:
Second to this request.

Once we have current reconciliation processing merged, the channel for notification of EoR (all) received from BGP to fpmsyncd is necessary for a production quality BGP docker warm reboot feature.

Reply:
@heidinet2007 Can you please elaborate on this: "because admins have alternative ways to achieve zero packet loss during routing stack upgrade".

I do realize that there's room for suboptimal routing (or even drops) in the interval comprehended between the bgp-peers sending their EoRs, and the warmrestart-timer kicking off, but this is expected to be a very short interval if the timers are properly set.

I have already started to prototype the EoR-to-fpmsyncd solution but that won't be ready for this coming release (only two weeks).

Reply:
EoR won't work even if routing application could send it since there is no notion of a table AF in sonic and all prefixes are in one bucket.

Reply:
It is not EoR from BGP to fpmsyncd directly,  the path will be BGP-->Zebra-->fpmsyncd. 
With proper processing zebra should make sure the EoR(zebra) is sent to fpmsyncd after all neighbors have PEER_STATUS_EOR_RECEIVED and routes pushed to fpmsyncd.

Reply:
@nikos-github Zooming out a little bit, the fundamental problem that we are trying to fix here is how to close the 'gap' between sonic's control and fwd plane to expedite the reconciliation process. So in my mind, given that SONiC is AFI/SAFI agnostic, there's no real need to treat each AFI/SAFI independently; as long as there's one missing EoR for a given AFI/SAFI, we know that no reconciliation process should take place (obviously, within a certain time boundary).

Of course, this imply that we will be penalizing faster AFI/SAFIs, but the alternative to this approach would be to introduce AFI/SAFI semantics in SONiC, which as you know, is a longer-term project.

And yep, i'm aware of EoR being used outside GR context, it's just a requirement.

Reply:
@jipanyang I believe you are confusing things. There is no notion of an implicit EoR through the expiration of update-delay. The update-delay timer (not processing) is a catch all mechanism for BGP that controls the maximum time BGP will wait after the first neighbor is established and until it starts calculating bestpaths and sending out updates. You can certainly implement an EoR message based on stopping or upon expiration of the update-delay timer and while it is simple, it would also be wrong since you are not guaranteed that bestpath selection is finished, RIB installation is done, etc. In other words sending an EoR to RIB when the update-delay timer expires or is stopped, can most certainly be premature.

@rodnymolina What you are describing can potentially lead you to wait until the update-delay timer expires. Per my comment above, this has a good chance it will be premature in those cases.

Reply:
It might be confusing to use EoR here, as it's more like a BGP term. What we need here is a RIB/ZEBRA CONVERGED signal to fpm. RIB convergence time has dependancy on routing protocols (here we consider only BGP), current BGP implementation already has a scheme to tell RIB when it receives EoR from all its neighbours. We should be able to utilize it.

Reply:
Agree, I vote for using something more generic, like FRR's existent EOIU terminology, which can be potentially extended to other protocols in the future too.